Fix leak of internal IDs registered during compound datatype conversion.

Allegedly fixes segfault after running h5py test suite.

https://github.com/h5py/h5py/issues/2419
https://github.com/HDFGroup/hdf5/pull/4459
--- src/H5T.c.orig	2024-04-15 14:47:31.000000000 -0500
+++ src/H5T.c	2024-05-09 00:36:32.000000000 -0500
@@ -1643,9 +1643,8 @@
     FUNC_ENTER_PACKAGE_NOERR
 
     assert(dt);
-    assert(dt->shared);
 
-    if (H5T_STATE_IMMUTABLE == dt->shared->state) {
+    if (dt->shared && (H5T_STATE_IMMUTABLE == dt->shared->state)) {
         dt->shared->state = H5T_STATE_RDONLY;
         (*n)++;
     } /* end if */
@@ -1862,7 +1861,6 @@
 
     /* Sanity check */
     assert(dt);
-    assert(dt->shared);
 
     /* If this datatype is VOL-managed (i.e.: has a VOL object),
      * close it through the VOL connector.
@@ -4162,10 +4160,10 @@
     FUNC_ENTER_NOAPI(FAIL)
 
     /* Sanity check */
-    assert(dt && dt->shared);
+    assert(dt);
 
     /* Clean up resources, depending on shared state */
-    if (dt->shared->state != H5T_STATE_OPEN) {
+    if (dt->shared && (dt->shared->state != H5T_STATE_OPEN)) {
         if (H5T__free(dt) < 0)
             HGOTO_ERROR(H5E_DATATYPE, H5E_CANTFREE, FAIL, "unable to free datatype");
 
@@ -4202,10 +4200,9 @@
 
     /* Sanity check */
     assert(dt);
-    assert(dt->shared);
 
     /* Named datatype cleanups */
-    if (dt->shared->state == H5T_STATE_OPEN) {
+    if (dt->shared && (dt->shared->state == H5T_STATE_OPEN)) {
         /* Decrement refcount count on open named datatype */
         dt->shared->fo_count--;
 
--- src/H5Tconv.c.orig	2024-04-15 14:47:31.000000000 -0500
+++ src/H5Tconv.c	2024-05-09 00:37:43.000000000 -0500
@@ -2241,6 +2241,7 @@
             if (need_ids) {
                 hid_t tid;
 
+                if (priv->src_memb_id[i] == H5I_INVALID_HID) {
                 if ((tid = H5I_register(H5I_DATATYPE, priv->src_memb[i], false)) < 0) {
                     H5T__conv_struct_free(priv);
                     cdata->priv = NULL;
@@ -2248,7 +2249,9 @@
                                 "can't register ID for source compound member datatype");
                 }
                 priv->src_memb_id[i] = tid;
+                }
 
+                if (priv->dst_memb_id[src2dst[i]] == H5I_INVALID_HID) {
                 if ((tid = H5I_register(H5I_DATATYPE, priv->dst_memb[src2dst[i]], false)) < 0) {
                     H5T__conv_struct_free(priv);
                     cdata->priv = NULL;
@@ -2256,6 +2259,7 @@
                                 "can't register ID for destination compound member datatype");
                 }
                 priv->dst_memb_id[src2dst[i]] = tid;
+                }
             }
         } /* end if */
     }     /* end for */
